### ISSUE TITLE
ACLs that allow only certain topics

### ISSUE THREAD

**Issue by heya-naohiro** [mochi-co/mqtt#258]
Hello, This is my first Issue on this project.

If I want to allow only certain topics and exclude others, can I use "github.com/mochi-co/mqtt/v2/hooks/auth" to implement this?
The following implementation seems to work at first glance, but fails one out of several times.
The following will be output as a debug message 
```DBG client failed allowed ACL check```

```
package main

import (
	"log"
	"os"
	"os/signal"
	"syscall"

	"github.com/mochi-co/mqtt/v2"
	"github.com/mochi-co/mqtt/v2/hooks/auth"
	"github.com/mochi-co/mqtt/v2/listeners"
	"github.com/rs/zerolog"
)

func main() {
	server := mqtt.New(nil)
	l := server.Log.Level(zerolog.DebugLevel)
	server.Log = &l

	rules := &auth.Ledger{
		Auth: auth.AuthRules{
			{Remote: "*", Allow: true},
		},
		ACL: auth.ACLRules{ 
			{
				Filters: auth.Filters{
					"sometopic": auth.ReadWrite,
					"#":         auth.Deny,
				},
			},
		},
	}

	err := server.AddHook(new(auth.Hook), &auth.Options{
		Ledger: rules,
	})
	if err != nil {
		log.Fatal(err)
	}

	tcp := listeners.NewTCP("t1", ":1885", nil)

	err = server.AddListener(tcp)
	if err != nil {
		log.Fatal(err)
	}

	sigs := make(chan os.Signal, 1)
	done := make(chan bool, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigs
		done <- true
	}()

	go func() {
		err = server.Serve()
		if err != nil {
			log.Fatal(err)
		}
	}()
	<-done
	server.Close()
}

```

# MQTT-Server Log

```
6:28AM INF added hook hook=auth-ledger
6:28AM INF loaded auth rules acl=1 authentication=1 hook=auth-ledger
6:28AM INF attached listener address=:1885 id=t1 protocol=tcp
6:28AM INF mochi mqtt starting version=2.2.15
6:28AM INF mochi mqtt server started
6:28AM DBG system event loop started
6:28AM DBG client disconnected client=mosq-G2QKl8k6D4DGn1Rml3 listener=t1 remote=127.0.0.1:52012
6:28AM DBG client disconnected client=mosq-bAAxEidbRTXGDKMRnn listener=t1 remote=127.0.0.1:52028
6:28AM DBG client disconnected client=mosq-ecxUQ5hQcquG7KCmaL listener=t1 remote=127.0.0.1:52066
6:28AM DBG client failed allowed ACL check client=mosq-0lQ8uhhRXR0BQYRK66 hook=auth-ledger topic=sometopic username= ←★
6:28AM DBG client disconnected client=mosq-0lQ8uhhRXR0BQYRK66 listener=t1 remote=127.0.0.1:52076
6:28AM DBG client disconnected client=mosq-lWdGhoqrSx8jP2LzAZ listener=t1 remote=127.0.0.1:52090
6:28AM DBG client disconnected client=mosq-ENPXX2u8PRn3cLtNqv listener=t1 remote=127.0.0.1:52096
6:28AM DBG client disconnected client=mosq-cD6P6XkeCUvu3ugHFm listener=t1 remote=127.0.0.1:52100
6:28AM DBG client disconnected client=mosq-neHGbpMMqJOITJYhwm listener=t1 remote=127.0.0.1:52114
6:28AM DBG client disconnected client=mosq-cAMX0jf8kLrN36Ubyq listener=t1 remote=127.0.0.1:52128
6:28AM DBG client disconnected client=mosq-2wcBGJxHJAerz9gnFa listener=t1 remote=127.0.0.1:50650
6:28AM DBG client disconnected client=mosq-YtKs7jHTTMZ5eOdxa9 listener=t1 remote=127.0.0.1:50656
6:28AM DBG client disconnected client=mosq-vXEkUKZeZaNw8l08lQ listener=t1 remote=127.0.0.1:50660
6:28AM DBG client disconnected client=mosq-6QZWBiE0Ka6cL8D2OX listener=t1 remote=127.0.0.1:50676
6:28AM DBG client disconnected client=mosq-SP75kzKupLo3baNctd listener=t1 remote=127.0.0.1:50686
6:28AM DBG client disconnected client=mosq-Ib7uIFh1ApfUftYDi0 listener=t1 remote=127.0.0.1:50696
6:28AM DBG client disconnected client=mosq-bBGs4kV91tnviLc7HZ listener=t1 remote=127.0.0.1:50710
6:28AM DBG client disconnected client=mosq-ENwPy35idhyCZCXb9w listener=t1 remote=127.0.0.1:50716
6:28AM DBG client disconnected client=mosq-HyluW6PlEJiu5SIT74 listener=t1 remote=127.0.0.1:50718
6:28AM DBG client failed allowed ACL check client=mosq-oicigjuXVscYqbKu0A hook=auth-ledger topic=sometopic username= ←★
6:28AM DBG client disconnected client=mosq-oicigjuXVscYqbKu0A listener=t1 remote=127.0.0.1:50730
6:28AM DBG client disconnected client=mosq-jfAf6XyE1aulMaLbPg listener=t1 remote=127.0.0.1:50746
6:28AM DBG client disconnected client=mosq-cchV0DXNBPXPz2Uc3n listener=t1 remote=127.0.0.1:50756
6:28AM DBG client disconnected client=mosq-t6cdxkp89AzK1s52Cm listener=t1 remote=127.0.0.1:50762
6:28AM DBG client failed allowed ACL check client=mosq-8nlW9igVJMZ7EnQCqA hook=auth-ledger topic=sometopic username= ←★
6:28AM DBG client disconnected client=mosq-8nlW9igVJMZ7EnQCqA listener=t1 remote=127.0.0.1:50776
```

# Client side 
I repeated  the following mosquitto command
```
$ mosquitto_pub -d -h localhost -p 1885 -t "sometopic" -m "hello world" -q 0
Client mosq-8nlW9igVJMZ7EnQCqA sending CONNECT
Client mosq-8nlW9igVJMZ7EnQCqA received CONNACK (0)
Client mosq-8nlW9igVJMZ7EnQCqA sending PUBLISH (d0, q0, r0, m1, 'sometopic', ... (11 bytes))
Client mosq-8nlW9igVJMZ7EnQCqA sending DISCONNECT
```

# Investigate 

I think this is because golang's "range" select randomly.

https://github.com/mochi-co/mqtt/blob/cb99d6f4bc2bce29e1de5f1f9fcf976d6c6e2824/hooks/auth/ledger.go#L191

Is there a better way to create an ACL if I only want to allow certain topics?
If anyone knows, please let me know.

**Comment by mochi-co** [comment]
Hi @heya-naohiro! Welcome and thanks for your issue :) 

I've checked and you are absolutely correct in your investigation. I've opened a PR with a fix in #262 👍🏻 Feel free to give it a review!

**Comment by heya-naohiro** [comment]
Thank you very much!
I have confirmed that it has been fixed.
